src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideStreamMemoryBasedAsyncOverrides.cs

```diff
@@ -0,0 +1,220 @@
+                //  if there are any bases in the middle.
+
+                if (!symbols.StreamType.Equals(type.BaseType, SymbolEqualityComparer.Default))
+                    return;
+
+                //  We use 'FirstOrDefault' because there could be multiple overrides for the same method if
+                //  there are compiler errors.
+
+                IMethodSymbol? readAsyncArrayOverride = GetOverridingMethodSymbols(type, symbols.ReadAsyncArrayMethod).FirstOrDefault();
+                IMethodSymbol? readAsyncMemoryOverride = GetOverridingMethodSymbols(type, symbols.ReadAsyncMemoryMethod).FirstOrDefault();
+                IMethodSymbol? writeAsyncArrayOverride = GetOverridingMethodSymbols(type, symbols.WriteAsyncArrayMethod).FirstOrDefault();
+                IMethodSymbol? writeAsyncMemoryOverride = GetOverridingMethodSymbols(type, symbols.WriteAsyncMemoryMethod).FirstOrDefault();
+
+                //  For both ReadAsync and WriteAsync, if the array-based form is overridden and the memory-based
+                //  form is not, we report a diagnostic. We report separate diagnostics for ReadAsync and WriteAsync. 
+
+                if (readAsyncArrayOverride is not null && readAsyncMemoryOverride is null)
+                {
+                    var diagnostic = CreateDiagnostic(type, readAsyncArrayOverride, symbols.ReadAsyncMemoryMethod);
+                    context.ReportDiagnostic(diagnostic);
+                }
+
+                if (writeAsyncArrayOverride is not null && writeAsyncMemoryOverride is null)
+                {
+                    var diagnostic = CreateDiagnostic(type, writeAsyncArrayOverride, symbols.WriteAsyncMemoryMethod);
+                    context.ReportDiagnostic(diagnostic);
+                }
+            }
+
+            static Diagnostic CreateDiagnostic(INamedTypeSymbol violatingType, IMethodSymbol arrayBasedOverride, IMethodSymbol memoryBasedMethod)
+            {
+                RoslynDebug.Assert(arrayBasedOverride.OverriddenMethod is not null);
+
+                //  We want to underline the name of the violating type in the class declaration. If the violating type
+                //  is a partial class, we underline all partial declarations.
+
+                return violatingType.CreateDiagnostic(
+                    Rule,
+                    violatingType.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                    arrayBasedOverride.OverriddenMethod.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                    memoryBasedMethod.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat));
+            }
+        }
+
+        private static bool TryGetRequiredSymbols(Compilation compilation, out RequiredSymbols requiredSymbols)
+        {
+            var int32Type = compilation.GetSpecialType(SpecialType.System_Int32);
+            var byteType = compilation.GetSpecialType(SpecialType.System_Byte);
```

Comment:
@mavasani Can these be null under any case?

Reply:
Yes, they can be null for broken compilation references. We should have a null check for these types as well.

Reply:
@mavasani Fixed. Sorry I missed this earlier.